apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/HiPerfSubscriber.cs

```diff
@@ -0,0 +1,723 @@
+﻿// Copyright 2017, Google Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Api.Gax;
+using Google.Api.Gax.Grpc;
+using Google.Apis.Auth.OAuth2;
+using Google.Cloud.PubSub.V1.Tasks;
+using Grpc.Auth;
+using Grpc.Core;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+// This class uses TaskHelper.ConfigureAwait, rather than directly calling .ConfigureAwait().
+// When running in a non-test environment this indirectly calls .ConfigureAwait(false).
+// Disable the ConfigureAwaitChecker warning:
+#pragma warning disable ConfigureAwaitChecker // CAC001
+
+namespace Google.Cloud.PubSub.V1
+{
+    /// <summary>
+    /// A PubSub subscriber that is associated with a specific <see cref="SubscriptionName"/>.
+    /// </summary>
+    /// <remarks>
+    /// <para>To receive messages, the <see cref="StartAsync"/> method must be called,
+    /// with a suitable message handler.</para>
+    /// <para>The message handler <see cref="Reply"/> states whether to acknowledge the message;
+    /// if acknowledged then (under normal conditions) it will not be received on this subscription again.</para>
+    /// <para>But note that it is always possible to receive duplicate messages. This services
```

Comment:
I wonder whether we're basically encouraging each user to build their own backlog of handled message IDs so that they can dedupe... could we do that instead?